core/main/src/broker/endpoint_broker.rs

```diff
@@ -0,0 +1,598 @@
+            app_id: value.app_id,
+        }
+    }
+}
+
+impl BrokerSender {
+    // Method to send the request to the underlying broker for handling.
+    pub async fn send(&self, request: BrokerRequest) -> RippleResponse {
+        if let Err(e) = self.sender.send(request).await {
+            error!("Error sending to broker {:?}", e);
+            Err(RippleError::SendFailure)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct EndpointBrokerState {
+    endpoint_map: Arc<RwLock<HashMap<String, BrokerSender>>>,
+    callback: BrokerCallback,
+    request_map: Arc<RwLock<HashMap<u64, BrokerRequest>>>,
+    extension_request_map: Arc<RwLock<HashMap<u64, ExtnMessage>>>,
+    rule_engine: RuleEngine,
+}
+
+impl EndpointBrokerState {
+    pub fn get(tx: Sender<BrokerOutput>, rule_engine: RuleEngine) -> Self {
+        Self {
+            endpoint_map: Arc::new(RwLock::new(HashMap::new())),
+            callback: BrokerCallback { sender: tx },
+            request_map: Arc::new(RwLock::new(HashMap::new())),
+            extension_request_map: Arc::new(RwLock::new(HashMap::new())),
+            rule_engine,
+        }
+    }
+
+    fn get_request(&self, id: u64) -> Result<BrokerRequest, RippleError> {
+        let result = { self.request_map.read().unwrap().get(&id).cloned() };
+        if result.is_none() {
+            return Err(RippleError::InvalidInput);
+        }
+
+        let result = result.unwrap();
+        if !result.rpc.is_subscription() {
+            let _ = self.request_map.write().unwrap().remove(&id);
+        }
+        Ok(result)
+    }
+
+    fn get_extn_message(&self, id: u64) -> Result<ExtnMessage, RippleError> {
+        let result = { self.extension_request_map.write().unwrap().remove(&id) };
+        match result {
+            Some(v) => Ok(v),
+            None => Err(RippleError::NotAvailable),
+        }
+    }
+
+    fn update_request(
+        &self,
+        rpc_request: &RpcRequest,
+        rule: Rule,
+        extn_message: Option<ExtnMessage>,
+    ) -> BrokerRequest {
+        ATOMIC_ID.fetch_add(1, Ordering::Relaxed);
+        let id = ATOMIC_ID.load(Ordering::Relaxed);
+        let mut rpc_request_c = rpc_request.clone();
+        {
+            let mut request_map = self.request_map.write().unwrap();
+            let _ = request_map.insert(
+                id,
+                BrokerRequest {
+                    rpc: rpc_request.clone(),
+                    rule: rule.clone(),
+                },
+            );
+        }
+
+        if extn_message.is_some() {
+            let mut extn_map = self.extension_request_map.write().unwrap();
+            let _ = extn_map.insert(id, extn_message.unwrap());
+        }
+
+        rpc_request_c.ctx.call_id = id;
+        self.get_broker_request(&rpc_request_c, rule)
+    }
+
+    pub fn build_thunder_endpoint(&mut self) {
+        if let Some(endpoint) = self.rule_engine.rules.endpoints.get("thunder").cloned() {
```

Comment:
Where are the keys like 'thunder" defined? Should they be listed explicitly via an enum or something?

Reply:
Yes, the new architecture expects Thunder to be default supported outside the SDK. Extensions for Thunder(and SDK) might not be needed in the future based on this new architecture